listeners.js

```diff
@@ -0,0 +1,156 @@
+const _ = require('lodash')
+const jsonpointer = require('jsonpointer')
+
+class Listener{
+    constructor(schema, ui_schema, formRef){
+        this.schema = Object.assign({}, schema)
+        this.ui_schema = Object.assign({}, ui_schema)
+        this.formRef = formRef
+        this.listeners = this._getListeners()
+    }
+
+    _getListeners() {
+        let acc = {}
+        this._getListenersFromUISchema(this.ui_schema, '', acc)
+        return acc
+    }
+
+    _getListenersFromUISchema(ui_schema, path, acc) {
+        const listeners = ui_schema["listener"] || ui_schema["logic:listener"]
+        if(listeners){
+            acc[path] = (acc[path] || [])
+            if (_.isArray(listeners)) {
+                acc[path].concat(listeners)
+            }
+            else{
+                acc[path].push(listeners)
+            }
+        }
+
+        if(ui_schema["fields"]) {
+            _.each(ui_schema["fields"], function (value, key) {
+                this._getListenersFromUISchema(value, path + '/' + key, acc)
+            }.bind(this))
+        }
+
+        else if (ui_schema["item"]) {
+            if (ui_schema["item"]["fields"]) {
+                _.each(ui_schema["item"]["fields"], function (value, key) {
+                    this._getListenersFromUISchema(value, path + '/' + key, acc)
+                }.bind(this))
+            }
+        }
+    }
+
+    update(value, path) {
+        let form = Object.assign({}, value)
+        // Remove integers from path which are array element pointers and
+        // unncessary for the path listeners.
+        let jsonPointerPath =
+            _.reduce(path,
+                function (acc, piece) {
+                    return _.isString(piece) ? `${acc}/${piece}` : acc;
+                },
+                '')
+        _.each(this.listeners, (listenerInfo, key) => {
+            // We want to check for the current key, but also any parents that are listening
+            if (_.startsWith(jsonPointerPath, key)) {
+                const formSubSection = jsonpointer.get(form, key)
+                const {schema, ui_schema} = this._getSchemaAtPath(key)
+                _.each(listenerInfo, (val) => this._fireListenerLogicAndSetState(val, formSubSection, schema, ui_schema))
+            }
+        })
+    }
+
+    _fireListenerLogicAndSetState = (val, storeSubSection, schema, ui_schema) => {
+        if(val["function"] && val["function"] == "list_adder") {
```

Comment:
This can be simplified by assigning `val["function"]` to a variable and exiting early if it's undefined. Then `this._maybeUpdateComponentStateInsideForm(val.output_key, output)` can be called outside the if/ else.